app/bundles/EmailBundle/Tests/Model/EmailActionModelTest.php

```diff
@@ -0,0 +1,136 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Mautic\EmailBundle\Tests\Model;
+
+use Mautic\CategoryBundle\Entity\Category;
+use Mautic\EmailBundle\Entity\Email;
+use Mautic\EmailBundle\Model\EmailActionModel;
+use Mautic\EmailBundle\Model\EmailModel;
+use PHPUnit\Framework\MockObject\MockObject;
+use PHPUnit\Framework\TestCase;
+
+class EmailActionModelTest extends TestCase
```

Comment:
I usually prefer to write both but I see your point. If the preferred way is to go with only e2e I can get onboard with that ;)

Reply:
Over the years I've learned that if you have a bunch of unit tests covering a feature then it doesn't mean the feature is working. If you refactor the feature, you have to refactor also the unit tests because they contain a lot of implementation details. So it adds work and it won't catch all the bugs.

On the other hand, if you have a functional test that is sending a specific input and asserting for specific output or database state then you are 99% sure that the feature works. And when you refactor the feature, you don't have to touch the tests and be sure that the feature still works.

Reply:
No problem. Do you think it'd be better to remove this class altogether?

Reply:
If it will be covered by a functional test for both scenarios where the user have the permission and doesn't have it then this unit test can be removed.

Reply:
I'm not sure about this. I usually try to keep my e2e tests to cover the happy path and deal with edge cases at a lower level. Mainly to prevent e2e suites to be too heavy/time consuming to run.

Reply:
Then a functional test in PHPUNIT would be the way to go.

Reply:
I'll see about that. Thanks

Reply:
Hey @escopecz I just checked the code and I noticed there was already a test `\Mautic\EmailBundle\Tests\Model\EmailActionModelTest::testDoesntSetNewCategoryForNonEditableEmails`, would this do?

Reply:
Can you link where it is? I can't find it

Reply:
Running a composer test locally, there are still some tests failing

Reply:
@mchojrin Hey, got some minutes to have a look?

Reply:
@mchojrin Please don't give up, we're almost there! 🙏

Reply:
Apologies for late replies. I got a bit distracted with work. I have several PRs on my todo list to re-review. I think I'll find a few hours today.